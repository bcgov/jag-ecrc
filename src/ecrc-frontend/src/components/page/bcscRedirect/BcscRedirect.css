.option {
  width: 100%;
  background-color: rgba(239, 243, 252, 1);
  box-sizing: border-box;
  border-width: 1px;
  border-style: solid;
  border-color: rgba(190, 206, 243, 1);
  border-radius: 6px;
  min-height: 70px;
  font-weight: 700;
  font-style: normal;
  font-size: 16px;
  color: #003366;
  text-align: left;
  margin-top: 20px;
  padding-left: 20px;
  line-height: 20px;
  padding-top: 15px;
  padding-bottom: 12px;
}

.option:hover {
  background-color: rgba(227, 234, 249, 1);
}

.alignRight {
  text-align: right;
}
