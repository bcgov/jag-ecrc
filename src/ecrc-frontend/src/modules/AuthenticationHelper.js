/* eslint-disable import/prefer-default-export */
import axios from "axios";
import { v4 as uuidv4 } from "uuid";

const jwt = require("jsonwebtoken");

export function isAuthenticated() {
  const token = sessionStorage.getItem("jwt");
  const validator = sessionStorage.getItem("validator");
  let isAuthed = true;

  if (!token || !validator) return false;

  // verify a token symmetric
  jwt.verify(token, validator, err => {
    if (err) {
      isAuthed = false;
    }
  });

  return isAuthed;
}

export function isAuthorized() {
  const token = sessionStorage.getItem("jwt");
  const validator = sessionStorage.getItem("validator");
  let isAuthorize = false;

  if (!token || !validator) return false;

  // verify a token symmetric
  jwt.verify(token, validator, (err, decoded) => {
    if (err || !decoded.authorities) {
      isAuthorize = false;
    } else if (decoded.authorities.includes("Authorized")) {
      isAuthorize = true;
    }
  });

  return isAuthorize;
}

export function isActionPerformed(action) {
  const token = sessionStorage.getItem("jwt");
  const validator = sessionStorage.getItem("validator");
  let isPerformed = false;

  if (!token || !validator) return false;

  // verify a token symmetric
  jwt.verify(token, validator, (err, decoded) => {
    if (err || !decoded.actionsPerformed) {
      isPerformed = false;
    } else if (decoded.actionsPerformed.includes(action)) {
      isPerformed = true;
    }
  });

  return isPerformed;
}

export function storeValidator() {
  const uuid = sessionStorage.getItem("uuid");

  if (!uuid) return false;

  axios
    .get(`/ecrc/initialHandshake?requestGuid=${uuid}`)
    .then(res => {
      const value = res.data;

      if (value) {
        sessionStorage.setItem("validator", value);
      }
    })
    .catch(() => {
      return false;
    });

  return true;
}

export function storeUUID() {
  const uuid = uuidv4();
  sessionStorage.setItem("uuid", uuid);
}

export function generateJWTToken(payload) {
  const validator = sessionStorage.getItem("validator");

  if (!validator) return false;

  let token;

  if (payload.exp) {
    token = jwt.sign(payload, validator);
  } else {
    token = jwt.sign(payload, validator, { expiresIn: "1h" });
  }

  sessionStorage.setItem("jwt", token);

  return token;
}

export function accessJWTToken(token) {
  const validator = sessionStorage.getItem("validator");

  const payload = jwt.verify(token, validator);

  return payload;
}
