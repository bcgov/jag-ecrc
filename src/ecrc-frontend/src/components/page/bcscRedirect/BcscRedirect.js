/* eslint-disable react/jsx-one-expression-per-line */
import React, { useState } from "react";
import PropTypes from "prop-types";
import { Redirect } from "react-router-dom";
import axios from "axios";

import Header from "../../base/header/Header";
import Footer from "../../base/footer/Footer";
import { Button } from "../../base/button/Button";
import SideCards from "../../composite/sideCards/SideCards";
import {
  isAuthenticated,
  isActionPerformed
} from "../../../modules/AuthenticationHelper";

import "../page.css";
import "./BcscRedirect.css";

export default function BcscRedirect({ page: { header, saveOrg, setError } }) {
  const [bcscUrl, setBcscUrl] = useState("");
  const [toError, setToError] = useState(false);

  React.useEffect(() => {
    window.scrollTo(0, 0);

    if (!isAuthenticated() || !isActionPerformed("tou")) {
      setError({
        status: 403
      });
      setToError(true);
    }

    const token = sessionStorage.getItem("jwt");
    const uuid = sessionStorage.getItem("uuid");

    axios
      .get(`/ecrc/protected/getBCSCUrl?requestGuid=${uuid}`, {
        headers: {
          Authorization: `Bearer ${token}`
        }
      })
      .then(res => {
        setBcscUrl(res.data);
      })
      .catch(error => {
        setToError(true);
        if (error && error.response && error.response.status) {
          setError({
            status: error.response.status,
            message: error.response.data
          });
        }
      });
  }, [setError]);

  const loginBtn = {
    label: "Login with a BC Services Card",
    buttonStyle: "btn ecrc_btn_login_bcsc",
    buttonSize: "btn",
    type: "button"
  };

  const onLoginClick = () => {
    if (!isAuthenticated()) {
      setError({
        status: 590,
        message: "Session Expired"
      });
      setToError(true);
      return;
    }
    sessionStorage.setItem("validExit", true);
    saveOrg();
    // REDIRECT TO BCSC
    window.open(bcscUrl, "_self");
  };

  if (toError) {
    return <Redirect to="/criminalrecordcheck/error" />;
  }

  return (
    <main>
      <Header header={header} />
      <div className="page">
        <div className="content col-md-8">
          <h1>Apply for a Criminal Record Check</h1>
          <p>
            To apply for a criminal record check online, you must use your{" "}
            <a
              href="https://www2.gov.bc.ca/gov/content/governments/government-id/bc-services-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              BC Services Card.
            </a>{" "}
          </p>
          <p>
            Only cards <b>with a photo</b> are accepted. If it&apos;s your first
            time using your BC Services Card to access an online service, you
            need to set up your account for use online. Completing a one-time
            security check to{" "}
            <a
              href="https://www2.gov.bc.ca/gov/content/governments/government-id/bc-services-card/login-with-card"
              target="_blank"
              rel="noopener noreferrer"
            >
              verify your identity
            </a>
            .
          </p>

          <p>Please select an option below:</p>
          <div className="option">
            <div className="row">
              <div className="col-sm-8 mt-lg-2 mb-2">
                I have a photo BC Services Card
              </div>
              <div className="col-sm-4">
                <Button button={loginBtn} onClick={onLoginClick} />
              </div>
            </div>
          </div>

          <div className="mt-5">
            <a href="/criminalrecordcheck/transition">
              I do not have a BC Services Card, or I have a non-photo BC
              Services Card
            </a>
            .
          </div>
        </div>

        <div className="sidecard">
          <SideCards type={"bcservice"} />
        </div>
      </div>
      <Footer />
    </main>
  );
}

BcscRedirect.propTypes = {
  page: PropTypes.shape({
    header: PropTypes.shape({
      name: PropTypes.string.isRequired
    }),
    saveOrg: PropTypes.func.isRequired,
    setError: PropTypes.func.isRequired
  }).isRequired
};
