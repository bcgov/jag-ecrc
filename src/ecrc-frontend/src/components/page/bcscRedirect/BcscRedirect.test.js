import React from "react";
import axios from "axios";
import { MemoryRouter, Router } from "react-router-dom";
import { render, fireEvent, getByText, wait } from "@testing-library/react";
import { createMemoryHistory } from "history";
import { generateJWTToken } from "../../../modules/AuthenticationHelper";

import BcscRedirect from "./BcscRedirect";

jest.mock("axios");

describe("BcscRedirect Page Component", () => {
  window.scrollTo = jest.fn();
  window.open = jest.fn().mockImplementation(() => true);

  const header = {
    name: "Criminal Record Check"
  };

  const saveOrg = jest.fn();
  const setError = jest.fn();

  const page = {
    header,
    saveOrg,
    setError
  };

  const axiosCall = {
    data: "http://test.com"
  };

  beforeEach(() => {
    sessionStorage.setItem("validator", "secret");
    sessionStorage.setItem("uuid", "unique123");
    generateJWTToken({
      actionsPerformed: ["tou"],
      authorities: ["Authorized"]
    });
  });

  test("Matches the snapshot", async () => {
    axios.get.mockResolvedValueOnce(axiosCall);
    const { asFragment } = render(
      <MemoryRouter>
        <BcscRedirect page={page} />
      </MemoryRouter>
    );
    await wait(() => {});
    expect(asFragment()).toMatchSnapshot();
  });

  test("Validate Login button", async () => {
    axios.get.mockResolvedValueOnce(axiosCall);
    const { container } = render(
      <MemoryRouter>
        <BcscRedirect page={page} />
      </MemoryRouter>
    );
    await wait(() => {});
    fireEvent.click(getByText(container, "Login with a BC Services Card"));
    expect(saveOrg).toHaveBeenCalled();
    expect(window.open).toHaveBeenCalledWith("http://test.com", "_self");
  });

  test("Validate login button click when session expired", async () => {
    axios.get.mockResolvedValueOnce(axiosCall);
    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <BcscRedirect page={page} />
      </Router>
    );
    await wait(() => {});

    sessionStorage.removeItem("jwt");

    fireEvent.click(getByText(container, "Login with a BC Services Card"));

    expect(setError).toHaveBeenCalled();
    expect(history.location.pathname).toEqual("/criminalrecordcheck/error");
  });

  test("Validate Redirect to Error when unauthorized", async () => {
    axios.get.mockResolvedValueOnce(axiosCall);
    const history = createMemoryHistory();
    generateJWTToken({
      actionsPerformed: ["none"],
      authorities: ["Authorized"]
    });
    render(
      <Router history={history}>
        <BcscRedirect page={page} />
      </Router>
    );
    await wait(() => {});
    expect(history.location.pathname).toEqual("/criminalrecordcheck/error");
  });

  test("Validate Redirect to Error page with error response/status", async () => {
    axios.get.mockRejectedValueOnce({ response: { status: 400 } });
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <BcscRedirect page={page} />
      </Router>
    );
    await wait(() => {});
    expect(setError).toHaveBeenCalled();
    expect(history.location.pathname).toEqual("/criminalrecordcheck/error");
  });

  test("Validate Redirect to Error page without error response/status", async () => {
    axios.get.mockRejectedValueOnce();
    const history = createMemoryHistory();
    render(
      <Router history={history}>
        <BcscRedirect page={page} />
      </Router>
    );
    await wait(() => {});
    expect(history.location.pathname).toEqual("/criminalrecordcheck/error");
  });
});
