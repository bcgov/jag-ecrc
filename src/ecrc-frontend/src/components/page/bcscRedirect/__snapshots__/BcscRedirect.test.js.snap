// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BcscRedirect Page Component Matches the snapshot 1`] = `
<main>
  <header>
    <nav
      className="navbar navbar-expand-lg navbar-dark"
    >
      <div
        className="container"
      >
        <a
          className="navbar-brand"
          href="/"
        >
          <img
            alt="B.C. Government Logo"
            className="img-fluid d-none d-md-block"
            height="44"
            src="/images/bcid-logo-rev-en.svg"
            width="181"
          />
          <img
            alt="B.C. Government Logo"
            className="img-fluid d-md-none"
            height="44"
            src="/images/bcid-symbol-rev.svg"
            width="64"
          />
        </a>
        <div>
          <a
            className="navbar-brand nav-item nav-link"
            href="/"
          >
            Criminal Record Check
          </a>
        </div>
        <button
          aria-controls="navbarNavAltMarkup"
          aria-expanded="false"
          aria-label="Toggle navigation"
          className="navbar-toggler"
          data-target="#navbarNavAltMarkup"
          data-toggle="collapse"
          type="button"
        >
          <span
            className="navbar-toggler-icon"
          />
        </button>
        <div
          className="collapse navbar-collapse"
          id="navbarNavAltMarkup"
        />
      </div>
    </nav>
  </header>
  <div
    className="page"
  >
    <div
      className="content col-md-8"
    >
      <h1>
        Apply for a Criminal Record Check
      </h1>
      <br />
      <p>
        To apply for a criminal record check online, you must use your
         
        <a
          href="https://www2.gov.bc.ca/gov/content/governments/government-id/bc-services-card/login-with-card"
        >
          BC Services Card
        </a>
         
        Account. Only cards 
        <b>
          with a photo
        </b>
         are accepted. If it's your first time using your BC Services Card to access an online service, you need to set up your account for use online. Completing a one-time security check to
         
        <a
          href="https://www2.gov.bc.ca/gov/content/governments/government-id/bc-services-card/login-with-card"
        >
          verify your identity
        </a>
        .
      </p>
      <p>
        Please select an option below:
      </p>
      <br />
      <div
        className="option"
      >
        <div
          className="row"
        >
          <div
            className="col-lg-6 col-md-12 "
            style={
              Object {
                "marginBottom": "10px",
              }
            }
          >
            I have a BC Services Card
             
            <span
              className="underlineText"
            >
              with a photo
            </span>
          </div>
          <div
            className="col-lg-6 col-md-4 alignRight"
          >
            <button
              className="btn ecrc_btn_login_bcsc btn"
              onClick={[Function]}
              type="button"
            >
              Login with your BC Services Card
              <div
                className="loader-hide"
              >
                <div
                  className="btn-loader"
                />
              </div>
            </button>
          </div>
        </div>
      </div>
      <div
        style={
          Object {
            "marginTop": "40px",
          }
        }
      >
        <a
          href="/criminalrecordcheck/transition"
        >
          I do not have a BC Services Card, or I have non-photo BC Services Card
        </a>
        .
      </div>
    </div>
    <div
      className="sidecard"
    >
      <div>
        <div
          className="wide-dashboard-spacing"
          style={
            Object {
              "position": "relative",
            }
          }
        >
          <div
            className="row"
          >
            <div
              className="col-12"
              style={
                Object {
                  "paddingTop": "30px",
                  "position": "relative",
                }
              }
            >
              <p />
              <section
                className="submit-container"
                id="blue-section"
              >
                <h2
                  className="heading-style"
                >
                  <div
                    className="row"
                  >
                    <div
                      className="round-icon-wrapper"
                    >
                      <svg
                        className="side-card-icon"
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        strokeWidth="0"
                        style={
                          Object {
                            "color": undefined,
                          }
                        }
                        viewBox="0 0 576 512"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M528 32H48C21.5 32 0 53.5 0 80v16h576V80c0-26.5-21.5-48-48-48zM0 432c0 26.5 21.5 48 48 48h480c26.5 0 48-21.5 48-48V128H0v304zm352-232c0-4.4 3.6-8 8-8h144c4.4 0 8 3.6 8 8v16c0 4.4-3.6 8-8 8H360c-4.4 0-8-3.6-8-8v-16zm0 64c0-4.4 3.6-8 8-8h144c4.4 0 8 3.6 8 8v16c0 4.4-3.6 8-8 8H360c-4.4 0-8-3.6-8-8v-16zm0 64c0-4.4 3.6-8 8-8h144c4.4 0 8 3.6 8 8v16c0 4.4-3.6 8-8 8H360c-4.4 0-8-3.6-8-8v-16zM176 192c35.3 0 64 28.7 64 64s-28.7 64-64 64-64-28.7-64-64 28.7-64 64-64zM67.1 396.2C75.5 370.5 99.6 352 128 352h8.2c12.3 5.1 25.7 8 39.8 8s27.6-2.9 39.8-8h8.2c28.4 0 52.5 18.5 60.9 44.2 3.2 9.9-5.2 19.8-15.6 19.8H82.7c-10.4 0-18.8-10-15.6-19.8z"
                        />
                      </svg>
                    </div>
                    <div
                      className="sideCardTitle"
                    >
                      Using the BC Services Card
                    </div>
                  </div>
                </h2>
                <div
                  className="submit-content"
                >
                  <span>
                    <div
                      className="side-card-text"
                    >
                      If you have a BC Services Card, you must use it to log in to the online security verification page.
                    </div>
                    <div>
                      <a
                        href="https://www2.gov.bc.ca/gov/content/governments/government-id/bc-services-card"
                      >
                        <div
                          className="side-card-link"
                        >
                          Learn more about the BC Services Card.
                        </div>
                      </a>
                    </div>
                    <div
                      className="side-card-text"
                    >
                      <br />
                      Eligibility for a BC Services Card is the same as for the Medical Services Plan(MSP).
                    </div>
                    <div>
                      <a
                        href="https://www2.gov.bc.ca/gov/content/health/health-drug-coverage/msp/bc-residents/eligibility-and-enrolment/are-you-eligible"
                      >
                        <div
                          className="side-card-link"
                        >
                          Learn more about BC Services Card eligibility.
                        </div>
                      </a>
                    </div>
                  </span>
                </div>
              </section>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <footer
    className="footer"
  >
    <nav
      className="navbar navbar-expand-sm navbar-dark"
    >
      <div
        className="container"
      >
        <ul
          className="navbar-nav"
        >
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/disclaimer"
              rel="noopener noreferrer"
              target="_blank"
            >
              Disclaimer
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/privacy"
              rel="noopener noreferrer"
              target="_blank"
            >
              Privacy
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/accessibility"
              rel="noopener noreferrer"
              target="_blank"
            >
              Accessibility
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/copyright"
              rel="noopener noreferrer"
              target="_blank"
            >
              Copyright
            </a>
          </li>
          <li
            className="nav-item"
          >
            <a
              className="nav-link"
              href="https://www2.gov.bc.ca/gov/content/home/contact-us"
              rel="noopener noreferrer"
              target="_blank"
            >
              Contact Us
            </a>
          </li>
        </ul>
      </div>
    </nav>
  </footer>
</main>
`;
